sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/storage/PreferenceTokenStorage.java

```diff
@@ -8,48 +8,103 @@
-     *
-     * @param context {@link Context} instance.
+     * Used for distinguishing users, string field for home account id.
      */
-    PreferenceTokenStorage(@NonNull Context context) {
-        mContext = context.getApplicationContext();
-    }
+    @VisibleForTesting
+    static final String PREFERENCE_KEY_HOME_ACCOUNT_ID = "AppCenter.home_account_id";
 
     /**
-     * Used for authentication requests, string field for auth token.
+     * Used for saving tokens history.
      */
     @VisibleForTesting
-    static final String PREFERENCE_KEY_AUTH_TOKEN = "AppCenter.auth_token";
+    static final String PREFERENCE_KEY_TOKEN_HISTORY = "AppCenter.auth_token_history";
 
     /**
-     * Used for distinguishing users, string field for home account id.
+     * The maximum number of tokens stored in the history.
      */
     @VisibleForTesting
-    static final String PREFERENCE_KEY_HOME_ACCOUNT_ID = "AppCenter.home_account_id";
+    static final int TOKEN_HISTORY_LIMIT = 5;
+
+    /**
+     * {@link Context} instance.
+     */
+    private final Context mContext;
+
+    /**
+     * Default constructor.
+     *
+     * @param context {@link Context} instance.
+     */
+    PreferenceTokenStorage(@NonNull Context context) {
+        mContext = context.getApplicationContext();
+    }
 
     @Override
-    public void saveToken(String token, String homeAccountId) {
-        AuthTokenContext.getInstance().setAuthToken(token, homeAccountId);
+    public void saveToken(String token, String homeAccountId, Date expiresOn) {
         String encryptedToken = CryptoUtils.getInstance(mContext).encrypt(token);
-        SharedPreferencesManager.putString(PREFERENCE_KEY_AUTH_TOKEN, encryptedToken);
-        SharedPreferencesManager.putString(PREFERENCE_KEY_HOME_ACCOUNT_ID, homeAccountId);
+        List<TokenStoreEntity> history = loadTokenHistory();
+        if (history == null) {
+            history = new ArrayList<TokenStoreEntity>() {{
+                add(new TokenStoreEntity(null, null, null));
+            }};
+        }
+
+        /* Do not add the same token twice in a row. */
+        String lastToken = history.size() > 0 ? history.get(history.size() - 1).getToken() : null;
+        if (lastToken != null && lastToken.length() > 0) {
+            CryptoUtils.DecryptedData decryptedData = CryptoUtils.getInstance(mContext).decrypt(lastToken, false);
+            lastToken = decryptedData.getDecryptedData();
+        }
+        if (!TextUtils.equals(lastToken, token)) {
+            history.add(new TokenStoreEntity(encryptedToken, new Date(), expiresOn));
+        }
+
+        /* Limit history size. */
+        if (history.size() > TOKEN_HISTORY_LIMIT) {
+            AppCenterLog.debug(LOG_TAG, "Size of the token history is exceeded. The oldest token has been removed.");
+            history.remove(0);
+        }
+
+        /* Update history and current token. */
+        saveTokenHistory(history);
+        if (token != null) {
```

Comment:
Why do we need to persist the current value on disk if history is on disk? Current can be in memory since all history is on disk.